### DISCUSSION TITLE
(Quasi)Static solve to generate initial fields with given traction separation law for cohesive zone model

### DISCUSSION THREAD

**Discussion by chunhuizhao478** [idaholab/moose#23139] [Q&A Modules: Solid mechanics] [answered]
Hi all, I'm trying to do a static solve (steady) or quasi-static solve (Transient, implicit) in order to generate the correct fields (displacements, stress etc for the domain) 

given the known boundary loading (use Neumann BC for specifying tractions and Dirichlet BC for fix corner points to remove rigid body motion . 

The geometry of the problem consists of two blocks separated by an interface (see the first attached picture), which is a cohesive zone interface quantified by a specific slip weakening law (tangential traction as a function of displacement jump, see the second attached picture).

I expect the stress field should be close the prescribed normal/tangent traction on the outer boundaries and along the interface some of the _qp points whose traction exceeds static friction strength would follow the slip weakening friction law and adjust their traction according (but most of the nodes should remain fixed). However, there seems to be a convergence /results not correct issue. I'm wondering if there is a better way to deal with the problem?

Below I attach the geometry & loading, slip weakening friction law, input file.

![IMG_EF1902E047DF-1](https://user-images.githubusercontent.com/93891863/211993464-2bc74ebb-6042-42fc-96e8-6a1f7b44723e.jpeg)

![IMG_9689C992D9AD-1](https://user-images.githubusercontent.com/93891863/211989099-7236a2f9-be54-4b27-816d-b300d7d0dc43.jpeg)

```
[Mesh]
    [./msh]
      type = GeneratedMeshGenerator
      dim = 2
      nx = 300
      ny = 300
      xmin = -15000
      xmax = 15000
      ymin = -15000
      ymax = 15000
    []
    [./new_block]
        type = ParsedSubdomainMeshGenerator
        input = msh
        combinatorial_geometry = 'y<0'
        block_id = 1
    []
    [./split]
        type = BreakMeshByBlockGenerator
        input = new_block
        split_interface = true
    []
    [interface]
        type = SideSetsBetweenSubdomainsGenerator
        input = split
        primary_block = 0
        paired_block = 1
        new_boundary = 'interface'
    []
    [secondary_interface]
        type = SideSetsBetweenSubdomainsGenerator
        input = interface
        primary_block = 1
        paired_block = 0
        new_boundary = 'secondary_interface'
    []
    [sideset1]
        type = SideSetsFromNodeSetsGenerator
        input = secondary_interface
    []
    [extranodeset2]
        type = ExtraNodesetGenerator
        coord = '-15000   -15000  0; 15000    15000  0;'
        new_boundary = 2ptrs
        input = sideset1
    []
    [extranodeset4]
        type = ExtraNodesetGenerator
        coord = '-15000  -15000  0;
                 -15000   15000  0;
                  15000  -15000  0;
                  15000   15000  0;'
        new_boundary = 4ptrs
        input = extranodeset2
    []
[]

[GlobalParams]
    displacements = 'disp_x disp_y'
[]

[Modules/TensorMechanics/CohesiveZoneMaster]
    [./czm_ik]
        boundary = 'Block0_Block1'
        strain = SMALL
        generate_output='traction_x traction_y jump_x jump_y normal_traction tangent_traction normal_jump tangent_jump'
    [../]
[]

[Modules]
    [./TensorMechanics]
        [./Master]
        [./all]
            strain = SMALL
            add_variables = true
            planar_formulation = PLANE_STRAIN
            generate_output = 'stress_xx stress_yy stress_xy'
            save_in = 'resid_x resid_y'
        [../]
        [../]
    [../]
[]

[Materials]
    [elasticity]
        type = ComputeIsotropicElasticityTensor
        lambda = 32.04e9
        shear_modulus = 32.04e9
        use_displaced_mesh = false
    []
    [stress]
        type = ComputeLinearElasticStress
    []
    [density]
        type = GenericConstantMaterial
        prop_names = density
        prop_values = 2670
    []
    [./czm_mat]
        type = SlipWeakeningTotalFormv2
        disp_slipweakening_x = disp_x
        disp_slipweakening_y = disp_y
        mu_d = mu_d
        mu_s = mu_s
        boundary = 'Block0_Block1'
    [../]
[]

[BCs]
    #outer boundaries
    [./fix_r_x]
        type = NeumannBC
        variable = disp_x
        boundary = right
        value = -120e6
    []
    [./fix_r_y]
        type = NeumannBC
        variable = disp_y
        boundary = right
        value = 70e6
    []
    [./fix_l_x]
        type = NeumannBC
        variable = disp_x
        boundary = left
        value = 120e6
    []  
    [./fix_l_y]
        type = NeumannBC
        variable = disp_y
        boundary = left
        value = -70e6
    []       
    [./fix_t_x]
        type = NeumannBC
        variable = disp_x
        boundary = top
        value = 70e6
    []
    [./fix_t_y]
        type = NeumannBC
        variable = disp_y
        boundary = top
        value = -120e6
    []    
    [./fix_b_x]
        type = NeumannBC
        variable = disp_x
        boundary = bottom
        value = -70e6
    []
    [./fix_b_y]
        type = NeumannBC
        variable = disp_y
        boundary = bottom
        value = 120e6
    []
    [./fix_ptrs_x]
        type = DirichletBC
        variable = disp_x
        boundary = 4ptrs
        value = 0
    []
    [./fix_ptrs_y]
        type = DirichletBC
        variable = disp_y
        boundary = 4ptrs
        value = 0
    []
[]

[Executioner]
    type = Transient
    solve_type = 'NEWTON'
    petsc_options_iname = '-pc_type -pc_hypre_type'
    petsc_options_value = ' hypre    boomeramg'
    nl_rel_tol = 1e-11
    nl_abs_tol = 1e-50
    num_steps = 1
[]

[Outputs]
    exodus = true
[]
```

**Comment by hugary1995** [comment]
~You can read the documentation on restart and recover, and also my comment in this PR: https://github.com/idaholab/moose/pull/21886#issuecomment-1251627435.~

**Reply by chunhuizhao478** [reply]
Hi Gary, thanks for the comment. Could you elaborate more about how this functionality is useful here? Do you mean the loading should be incremental buildup with restart?

**Reply by hugary1995** [reply, marked as answer]
I'm sorry I didn't read your question correctly. I suspect the convergence issue is coming from the traction-separation law. How do you define the traction-separation law when [[u]] is negative?

**Reply by hugary1995** [reply]
The easiest way for me to help is to setup a repository which I can pull down.

**Reply by chunhuizhao478** [reply]
Hi Gary, I think you make a good point, I don't know how to enforce [[u]] to be positive (no penetration is allowed) in czm, I do see there are some support in the contact module in the normal direction. I share this repo here if you want to take a look:  https://github.com/chunhuizhao478/czm_static_solve.

**Reply by hugary1995** [reply]
Thanks for setting up the repo. The way you obtain the local normal stress and shear stress doesn't seem quite right to me. You'll want to rotate the stress tensor first. You are also providing the wrong Jacobian, e.g., `_dinterface_traction_djump`. A correct Jacobian is critical to convergence. My suggestion for you right now is to first jot down the set of equations for the traction-separation law (and the Jacobian if you know how to derive it). This is helpful in many ways: 1. It will help me better understand your constitutive model. 2. It will be easier to derive the Jacobian. 3. You can later write code strictly following your equations.

**Reply by chunhuizhao478** [reply]
Much Thanks for your help! Sure, I will provide more information on the traction separation law by the end of this weekend.

**Reply by chunhuizhao478** [reply]
Hi, I have update the reference in `czm_static_solve/reference` folder, I also modify the computation for local normal stress and shear stress, however I don't know how to come up with the correct Jacobin since the algorithm contains if-else condition. The repo is re-phrased below: https://github.com/chunhuizhao478/czm_static_solve. Thanks for your time!

**Reply by GiudGiud** [reply]
Any update on this?
If the jacobian is wrong you should not use NEWTON for the solve. PJFNK will be more resilient to a bad jacobian
